src/provider/model/AppDirectory.ts

```
@@ -1,42 +1,118 @@
-        if (!this._directory) {
-            // If using the demo app directory in production, return an empty array.
-            if (process.env.NODE_ENV === 'production') {
-                this._directory = [];
-            } else {
-                const response: Response = await fetch('http://localhost:3923/provider/sample-app-directory.json');
-                if (response.ok) {
-                    this._directory = await response.json();
-                } else {
-                    throw new Error('Error fetching app directory: ' + response.statusText);
+    /**
+     * Fetch the AppDirectory.
+     * @param url Location of the AppDirectory.
+     */
+    private async fetchData(url: string): Promise<Application[]> {
+        const response: Response = await fetch(url);
+        if (response.ok) {
+            return response.json();
+        } else {
+            throw new Error('Error fetching app directory: ' + response.statusText);
+        }
+    }
+
+    /**
+     * Refresh the AppDirectory.
+     */
+    private async refreshDirectory(): Promise<void> {
+        // If using the demo app directory in production, set an empty array.
+        if (process.env.NODE_ENV === 'production') {
+            this.updateDirectory([]);
+            return;
+        }
+        const storedURL = await localForage.getItem<string>('url');
+        const currentURL = this._URL;
+        // Only update the directory if the URL has changed
+        // or the directory is empty (the last request may have failed?)
+        if (currentURL !== storedURL || (currentURL === storedURL && this._directory.length === 0)) {
+            try {
+                const applications = await this.fetchData(currentURL);
+                await this.updateDirectory(applications);
+                if (currentURL !== storedURL) {
+                    await this.updateURL(currentURL);
                 }
+            } catch (error) {
+                // Dont update
+                return;
```

What errors are we expecting here, any reason not to propagate upward? Completely ignoring an error like this is a bit scary-looking... :-)